protocols/upnp/src/provider.rs

```diff
@@ -0,0 +1,173 @@
+                let gateway = $gateway::search_gateway(options).await?;
+                let addr = gateway.get_external_ip().await?;
+
+                let (events_sender, mut task_receiver) = mpsc::channel(5);
+                let (mut task_sender, events_queue) = mpsc::channel(0);
+
+                $executor::spawn(async move {
+                    let mut futs = FuturesUnordered::new();
+                    loop {
+                        select! {
+                            req = task_receiver.select_next_some() => {
+                                let fut = match req {
+                                    GatewayRequest::AddMapping(mapping, duration) => {
+                                        let duration = duration.unwrap_or(0);
+                                        let gateway = gateway.clone();
+                                        async move {
+                                            match gateway
+                                                .add_port(
+                                                    mapping.protocol.into(),
+                                                    mapping.internal_addr.port(),
+                                                    mapping.internal_addr,
+                                                    duration,
+                                                    "rust-libp2p mapping",
+                                                ).await {
+                                                    Ok(()) => GatewayEvent::Mapped(mapping),
+                                                    Err(err) => GatewayEvent::MapFailure(mapping, err.into()),
+                                                }
+                                        }.boxed()
+                                    }
+                                    GatewayRequest::RemoveMapping(mapping) => {
+                                        let gateway = gateway.clone();
+                                        async move {
+                                            match gateway
+                                                .remove_port(
+                                                    mapping.protocol.into(),
+                                                    mapping.internal_addr.port()
+                                                ).await {
+                                                    Ok(()) => GatewayEvent::Removed(mapping),
+                                                    Err(err) => GatewayEvent::RemovalFailure(mapping, err.into()),
+                                                }
+                                        }.boxed()
+                                    }
+
+                                };
+                                futs.push(fut);
+                            },
+                            event = futs.select_next_some() => {
+                                task_sender.send(event).await.expect("receiver should be available");
+                            }
```

Comment:
Before accepting new work, i.e. polling `task_receiver.select_next_some()`, I suggest trying to finish existing work, i.e. `futs.select_next_some()`. See also https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#prioritize-local-work-over-new-work-from-a-remote.